File: php/src/Model/PushSubscriptionInput.php
<?php

declare(strict_types=1);

namespace Fivexer\SDK\Model;

/**
 * A browser Web Push subscription, shaped as `PushSubscription.toJSON()` returns it.
 *
 * The wire nests the two keys under a `keys` object while the browser hands them over flat, so
 * this builds its body explicitly rather than mapping property names.
 */
final class PushSubscriptionInput
{
    public function __construct(
        private readonly string $endpoint,
        private readonly string $p256dh,
        private readonly string $auth,
    ) {
    }

    /**
     * Serialise for the wire, nesting the keys the way the API expects.
     *
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        return [
            'endpoint' => $this->endpoint,
            'keys' => [
                'p256dh' => $this->p256dh,
                'auth' => $this->auth,
            ],
        ];
    }
}

File: java/src/main/java/io/fivexer/sdk/model/PushSubscriptionInput.java
package io.fivexer.sdk.model;

import com.google.gson.JsonObject;

/**
 * A browser Web Push subscription, shaped as {@code PushSubscription.toJSON()} returns it.
 *
 * <p>The wire nests the two keys under a {@code keys} object while the browser hands them over
 * flat, so this builds its body by hand rather than reflecting the field names.
 */
public class PushSubscriptionInput {
    private final String endpoint;
    private final String p256dh;
    private final String auth;

    public PushSubscriptionInput(String endpoint, String p256dh, String auth) {
        if (endpoint == null) {
            throw new IllegalArgumentException("endpoint is required");
        }
        if (p256dh == null) {
            throw new IllegalArgumentException("p256dh is required");
        }
        if (auth == null) {
            throw new IllegalArgumentException("auth is required");
        }
        this.endpoint = endpoint;
        this.p256dh = p256dh;
        this.auth = auth;
    }

    public String getEndpoint() { return endpoint; }
    public String getP256dh() { return p256dh; }
    public String getAuth() { return auth; }

    /** Serialise for the wire, nesting the keys the way the API expects. */
    public String toJson() {
        JsonObject keys = new JsonObject();
        keys.addProperty("p256dh", p256dh);
        keys.addProperty("auth", auth);
        JsonObject o = new JsonObject();
        o.addProperty("endpoint", endpoint);
        o.add("keys", keys);
        return o.toString();
    }
}
